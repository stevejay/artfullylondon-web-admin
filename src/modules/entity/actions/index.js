export const types = {
  GET_ENTITY: 'entity/GET_ENTITY',
  GET_ENTITY_STARTED: 'entity/GET_ENTITY_STARTED',
  GET_ENTITY_SUCCEEDED: 'entity/GET_ENTITY_SUCCEEDED',
  GET_ENTITY_FAILED: 'entity/GET_ENTITY_FAILED',
  SAVE_ENTITY: 'entity/SAVE_ENTITY',
  RESET_ENTITY_FOR_CREATE: 'entity/RESET_ENTITY_FOR_CREATE',
  CLEAR_ENTITY: 'entity/CLEAR_ENTITY'
  // UPDATE_ENTITY: 'entity/UPDATE_ENTITY',
  // AUTOCOMPLETE_SEARCH: 'entity/AUTOCOMPLETE_SEARCH',
  // AUTOCOMPLETE_SEARCH_SUCCEEDED: 'entity/AUTOCOMPLETE_SEARCH_SUCCEEDED',
  // GET_SUB_ENTITY: 'entity/GET_SUB_ENTITY',
  // GET_SUB_ENTITY_STARTED: 'entity/GET_SUB_ENTITY_STARTED',
  // GET_SUB_ENTITY_FINISHED: 'entity/GET_SUB_ENTITY_FINISHED',
  // CREATE_TALENT_FOR_EVENT: 'entity/CREATE_TALENT_FOR_EVENT',
  // GET_EVENT_AS_COPY: 'entity/GET_EVENT_AS_COPY',
  // GET_EVENT_AS_COPY_STARTED: 'entity/GET_EVENT_AS_COPY_STARTED',
  // GET_EVENT_AS_COPY_SUCCEEDED: 'entity/GET_EVENT_AS_COPY_SUCCEEDED',
  // GET_EVENT_AS_COPY_FAILED: 'entity/GET_EVENT_AS_COPY_FAILED',
  // TALENT_SELECTED: 'entity/TALENT_SELECTED'
}

export const getEntity = (entityType, id) => ({
  type: types.GET_ENTITY,
  payload: { entityType, id }
})

export const getEntityStarted = (entityType, id) => ({
  type: types.GET_ENTITY_STARTED,
  payload: { entityType, id }
})

export const getEntitySucceeded = (entityType, entity) => ({
  type: types.GET_ENTITY_SUCCEEDED,
  payload: { entityType, entity }
})

export const getEntityFailed = (entityType, statusCode) => ({
  type: types.GET_ENTITY_FAILED,
  payload: { entityType, statusCode: statusCode || 500 }
})

export const saveEntity = (
  entityType,
  values,
  isEdit,
  formName,
  normaliser,
  constraint,
  mapper
) => ({
  type: types.SAVE_ENTITY,
  payload: {
    entityType,
    values,
    isEdit,
    formName,
    normaliser,
    constraint,
    mapper
  }
})

export const resetEntityForCreate = entityType => ({
  type: types.RESET_ENTITY_FOR_CREATE,
  payload: { entityType }
})

export const clearEntity = entityType => ({
  type: types.CLEAR_ENTITY,
  payload: { entityType }
})
