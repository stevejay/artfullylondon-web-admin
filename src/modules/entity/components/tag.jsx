// @flow

import * as React from "react";
import { Text } from "grommet";
import AnimationBox from "shared/animation-box";

const BOX_PAD = { vertical: "hair", horizontal: "small" };
const BOX_MARGIN = { right: "xsmall", bottom: "xsmall" };

type Props = {
  +label: string
};

const Tag = ({ label }: Props) => (
  <AnimationBox
    animation="fadeIn"
    tag="span"
    direction="row"
    align="center"
    background="light-3"
    flex="shrink"
    justify="between"
    margin={BOX_MARGIN}
    pad={BOX_PAD}
    responsive={false}
    round="large"
  >
    <Text>{label}</Text>
  </AnimationBox>
);

export default Tag;
