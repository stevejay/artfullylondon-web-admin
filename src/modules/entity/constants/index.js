import PropTypes from 'prop-types'

import { FullVenue } from '_src/domain/venue'
import { FullTalent } from '_src/domain/talent'
import { FullEvent } from '_src/domain/event'
import { FullEventSeries } from '_src/domain/event-series'

export const EDITABLE_ENTITY_PROP_TYPE = PropTypes.oneOfType([
  PropTypes.instanceOf(FullTalent),
  PropTypes.instanceOf(FullVenue),
  PropTypes.instanceOf(FullEvent),
  PropTypes.instanceOf(FullEventSeries)
])
