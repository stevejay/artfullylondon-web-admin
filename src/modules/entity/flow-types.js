// @flow

export type Link = { +type: string, +url: string };
export type LinkArray = Array<Link>;

export type Image = {
  +id: string,
  +dominantColor: string,
  +copyright?: string
};
export type ImageArray = Array<Image>;

export type Tag = { +id: string, +label: string };
export type TagArray = Array<Tag>;

export type Review = {
  +source: string,
  +rating: number
};

export type EventTalent = {
  +roles: ?Array<string>,
  +characters: ?Array<string>,
  +id: string,
  +talent: {
    +id: string,
    +firstNames: ?string,
    +lastName: string,
    +images: ?ImageArray
  }
};

export type OpeningTime = {
  +from?: string,
  +to?: string
};

export type DayOpeningTime = {
  +day: number,
  +from: string,
  +to: string,
  +timesRangeId: string
};

export type DateOpeningTime = {
  +date: string,
  +from: string,
  +to: string
};

export type DateClosedTimeRange = {
  +date: string,
  +from: ?string,
  +to: ?string
};

export type TimesRange = {
  +id: string,
  +dateFrom: string,
  +dateTo: string,
  +label: string
};

export type Performance = {
  +at?: string
};

export type DayPerformance = {
  +day: number,
  +at: string,
  +timesRangeId: ?string
};

export type DatePerformance = {
  +date: string,
  +at: string
};

export type SpecialPerformance = {
  +date: string,
  +at: string,
  +audienceTags?: ?TagArray
};

export type DateClosedTimeAt = {
  +date: string,
  +at: ?string
};

export type SpecialOpeningTime = {
  +date: string,
  +from: string,
  +to: string,
  +audienceTags?: ?TagArray
};

export type EventForEdit = {
  +id: string,
  +status: string,
  +version: number,
  +links: ?LinkArray,
  +images: ?ImageArray,
  +weSay: ?string,
  +notes: ?string,
  +description: ?string,
  +descriptionCredit: ?string,
  +name: string,
  +eventType: string,
  +occurrenceType: string,
  +dateFrom: ?string,
  +dateTo: ?string,
  +costType: string,
  +costFrom: ?number,
  +costTo: ?number,
  +summary: string,
  +soldOut: ?boolean,
  +rating: number,
  +bookingType: string,
  +bookingOpens: ?string,
  +venueGuidance: ?string,
  +timedEntry: ?boolean,
  +duration: ?string,
  +minAge: ?number,
  +maxAge: ?number,
  +useVenueOpeningTimes: boolean,
  +timesRanges: ?Array<TimesRange>,
  +openingTimes: ?Array<DayOpeningTime>,
  +additionalOpeningTimes: ?Array<DateOpeningTime>,
  +specialOpeningTimes: ?Array<SpecialOpeningTime>,
  +openingTimesClosures: ?Array<DateClosedTimeRange>,
  +performances: ?Array<DayPerformance>,
  +additionalPerformances: ?Array<DatePerformance>,
  +specialPerformances: ?Array<SpecialPerformance>,
  +performancesClosures: ?Array<DateClosedTimeAt>,
  +soldOutPerformances: ?Array<DatePerformance>,
  +audienceTags: ?TagArray,
  +mediumTags: ?TagArray,
  +styleTags: ?TagArray,
  +geoTags: ?TagArray,
  +reviews: ?Array<Review>,
  +talents: ?Array<EventTalent>,
  +venueId: string,
  +venue: {
    +id: string,
    +name: string
  },
  +eventSeriesId: ?string,
  +eventSeries: ?{
    +id: string,
    +name: string
  }
};

export type EventSeriesForEdit = {
  +id: string,
  +status: string,
  +version: number,
  +links: ?LinkArray,
  +images: ?ImageArray,
  +weSay: ?string,
  +notes: ?string,
  +description: ?string,
  +descriptionCredit: ?string,
  +name: string,
  +eventSeriesType: string,
  +occurrence: string,
  +summary: string
};

export type TalentForEdit = {
  +id: string,
  +status: string,
  +version: number,
  +links: ?LinkArray,
  +images: ?ImageArray,
  +weSay: ?string,
  +notes: ?string,
  +description: ?string,
  +descriptionCredit: ?string,
  +firstNames: ?string,
  +lastName: string,
  +talentType: string,
  +commonRole: string
};

export type VenueForEdit = {
  +id: string,
  +status: string,
  +version: number,
  +links: ?LinkArray,
  +images: ?ImageArray,
  +weSay: ?string,
  +notes: ?string,
  +description: ?string,
  +descriptionCredit: ?string,
  +name: string,
  +venueType: string,
  +address: string,
  +postcode: string,
  +telephone: ?string,
  +email: ?string,
  +latitude: number,
  +longitude: number,
  +wheelchairAccessType: string,
  +disabledBathroomType: string,
  +hearingFacilitiesType: string,
  +hasPermanentCollection: boolean,
  +openingTimes: ?Array<DayOpeningTime>,
  +additionalOpeningTimes: ?Array<DateOpeningTime>,
  +openingTimesClosures: ?Array<DateClosedTimeRange>,
  +namedClosures: ?Array<string>
};
