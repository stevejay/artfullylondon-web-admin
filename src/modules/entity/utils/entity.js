// @flow

import type { ImageArray } from "../flow-types";

import isEmpty from "lodash/isEmpty";
import _ from "lodash";
import * as bookingType from "../types/booking-type";
import * as costType from "../types/cost-type";
import * as eventType from "../types/event-type";
import * as occurrenceType from "../types/occurrence-type";
import { getDateDisplayValue, getDurationAsTimeDisplayValue } from "./time";

export function getEnumDisplayValue(value: ?string): string {
  return value ? _.startCase(_.lowerCase(value.replace("_", " "))) : "";
}

export function tryGetMainImage(images: ?ImageArray) {
  return isEmpty(images) || !images ? null : images[0];
}

export function getTalentDisplayName(firstNames: ?string, lastName: string) {
  return `${firstNames ? firstNames + " " : ""}${lastName}`;
}

export function getOccurrenceDisplayValue(
  type: string,
  dateFrom: ?string,
  dateTo: ?string,
  event: string
) {
  switch (type) {
    case occurrenceType.ONE_TIME:
      return getDateDisplayValue(dateFrom);
    case occurrenceType.BOUNDED:
      return dateFrom === dateTo
        ? `On ${getDateDisplayValue(dateFrom)}`
        : `From ${getDateDisplayValue(dateFrom)} to ${getDateDisplayValue(
            dateTo
          )}`;
    default:
      return event === eventType.PERFORMANCE
        ? "Ongoing performances"
        : "Permanent exhibition";
  }
}

export function getCostDisplayValue(
  type: string,
  costFrom: ?number,
  costTo: ?number
) {
  if (type === costType.PAID) {
    const from = costFrom != null ? formatCost(costFrom) : "";
    const to = costTo != null ? formatCost(costTo) : "";
    return from === to
      ? `Tickets are ${from}`
      : `Tickets cost ${from} to ${to}`;
  } else {
    return "Free event";
  }
}

function formatCost(cost: number) {
  return cost === 0 ? "free" : `Â£${cost.toFixed(2)}`;
}

export function getBookingDisplayValue(
  type: string,
  soldOut: ?boolean,
  bookingOpens: ?string
) {
  if (type === bookingType.NOT_REQUIRED) {
    return "Booking is not required";
  }

  if (soldOut) {
    return "This event has sold out but tickets may still be available on the day";
  }

  const bookingValue =
    type === bookingType.REQUIRED
      ? "Booking is required"
      : "Booking is required for non-members";

  return bookingOpens
    ? `${bookingValue} and opens on ${getDateDisplayValue(bookingOpens)}`
    : bookingValue;
}

export function getDescriptionDisplayHtml(
  description: ?string,
  descriptionCredit: ?string
) {
  if (!description) {
    return "<p><em>No description available.</em></p>";
  } else if (!descriptionCredit) {
    return description;
  } else {
    return description.replace(
      /<\/p>$/,
      ` <em>(Information by ${descriptionCredit}.)</em></p>`
    );
  }
}

export function getDurationDisplayValue(duration: ?string, event: string) {
  if (!duration) {
    return null;
  }
  const timeValue = getDurationAsTimeDisplayValue(duration);
  return event === eventType.PERFORMANCE
    ? `Duration around ${timeValue}`
    : `Allow around ${timeValue} for your visit`;
}

export function getAgeDisplayValue(minAge: ?number, maxAge: ?number) {
  if (minAge && maxAge) {
    return `${minAge} to ${maxAge} years`;
  } else if (!maxAge && minAge) {
    return `${minAge} years and over`;
  } else if (!minAge && maxAge) {
    return `To ${maxAge} years`;
  } else {
    return null;
  }
}
