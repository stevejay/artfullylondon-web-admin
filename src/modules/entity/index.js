import EntityPage from './pages/entity'

import EntityFeaturedDetail from './components/featured-detail'
import EntityDescription from './components/description'
import EntitySectionHeading from './components/section-heading'
import EntityDetailsContainer from './components/details-container'
import EntityHeading from './components/heading'
import EntityWeSay from './components/we-say'
import EntityInfoBar from './components/info-bar'
import EntityColumnLayout from './components/column-layout'
import EntityColumn from './components/column'
import EntityAddress from './components/address'
import EntityDisabilityAccess from './components/disability-access'
import OpeningTimes from './components/opening-times'

import { moduleName, reducer, selectors } from './reducers'
import sagas from './sagas'
import * as actions from './actions'
import * as entityMapper from './lib/mapper'

export {
  EntityPage,
  EntityFeaturedDetail,
  EntityDescription,
  EntitySectionHeading,
  EntityDetailsContainer,
  EntityHeading,
  EntityWeSay,
  EntityInfoBar,
  EntityColumnLayout,
  EntityColumn,
  EntityAddress,
  EntityDisabilityAccess,
  OpeningTimes,
  entityMapper,
  moduleName,
  reducer,
  selectors,
  sagas,
  actions
}
