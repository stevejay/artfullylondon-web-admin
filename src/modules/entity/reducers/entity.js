import { handleActions } from 'redux-actions'
import RichTextEditor from 'react-rte'

import { FullEvent } from '_src/entities/event'
import { FullEventSeries } from '_src/entities/event-series'
import { FullTalent } from '_src/entities/talent'
import { FullVenue } from '_src/entities/venue'
import { types } from '_src/modules/entity/actions'
import * as entityConstants from '_src/constants/entity'
import { DEFAULT_MAP_CENTER } from '_src/modules/location'

const initialState = {
  entityId: null,
  entity: null,
  getInProgress: false,
  getFailed: false
}

export const reducer = handleActions(
  {
    [types.CLEAR_ENTITY]: () => initialState,
    [types.RESET_ENTITY_FOR_CREATE]: (state, action) => ({
      ...initialState,
      entity: createEntity(action.payload.entityType)
    }),
    [types.GET_ENTITY_STARTED]: (state, action) => ({
      ...state,
      entityId: action.payload.id,
      entity: null,
      getInProgress: true,
      getFailed: false
    }),
    [types.GET_ENTITY_SUCCEEDED]: (state, action) => ({
      ...state,
      getInProgress: false,
      entity: createEntity(action.payload.entityType, action.payload.entity)
    }),
    [types.GET_ENTITY_FAILED]: (state, action) => ({
      ...initialState,
      getFailed: true
    })
  },
  initialState
)

export const selectors = {
  entity: state => state.entity,
  entityId: state => state.entityId,
  gettingEntity: state => state.getInProgress,
  failedToGetEntity: state => state.getFailed
}

function createEntity (entityType, entity = null) {
  switch (entityType) {
    case entityConstants.ENTITY_TYPE_EVENT:
      return new FullEvent(entity)
    case entityConstants.ENTITY_TYPE_EVENT_SERIES:
      return new FullEventSeries(entity)
    case entityConstants.ENTITY_TYPE_TALENT:
      return new FullTalent(entity)
    case entityConstants.ENTITY_TYPE_VENUE:
      return new FullVenue(entity)
    default:
      throw new Error(`entityType ${entityType} not supported`)
  }
}

// TODO rip all this stuff out and put it in the domain!!
function getEntityDefaultValues (entityType) {
  switch (entityType) {
    // case entityConstants.ENTITY_TYPE_TALENT:
    //   return {
    //     id: null,
    //     status: entityConstants.ACTIVE_STATUS,
    //     validStatuses: [entityConstants.ACTIVE_STATUS],
    //     firstNames: '',
    //     lastName: '',
    //     talentType: talentConstants.TALENT_TYPE_INDIVIDUAL,
    //     commonRole: '',
    //     description: RichTextEditor.createEmptyValue(),
    //     links: [],
    //     images: [],
    //     currentEvents: [],
    //     futureEvents: [],
    //     weSay: '',
    //     version: 0,
    //     createdDate: null
    //   }
    case entityConstants.ENTITY_TYPE_VENUE:
      return {
        id: null,
        status: entityConstants.ACTIVE_STATUS,
        validStatuses: [entityConstants.ACTIVE_STATUS],
        name: '',
        venueType: '',
        address: '',
        pin: {
          lat: null,
          lng: null
        },
        defaultCenter: DEFAULT_MAP_CENTER,
        postcode: '',
        email: '',
        telephone: '',
        description: RichTextEditor.createEmptyValue(),
        wheelchairAccessType: '',
        disabledBathroomType: '',
        hearingFacilitiesType: '',
        hasPermanentCollection: false,
        openingTimes: [],
        additionalOpeningTimes: [],
        openingTimesClosures: [],
        namedClosures: '',
        links: [],
        images: [],
        weSay: '',
        notes: '',
        version: 0,
        createdDate: null
      }
    case entityConstants.ENTITY_TYPE_EVENT:
      return {
        id: null,
        status: entityConstants.ACTIVE_STATUS,
        validStatuses: [entityConstants.ACTIVE_STATUS],
        name: '',
        eventType: '',
        dateFrom: null,
        dateTo: null,
        costType: '',
        summary: '',
        description: RichTextEditor.createEmptyValue(),
        descriptionCredit: '',
        links: [],
        costFrom: '',
        costTo: '',
        bookingType: '',
        bookingOpens: null,
        venue: {},
        venueGuidance: '',
        eventSeries: {},
        duration: '',
        weSay: '',
        rating: '3',
        minAge: '',
        useVenueOpeningTimes: false,
        openingTimes: [],
        additionalOpeningTimes: [],
        specialOpeningTimes: [],
        openingTimesClosures: [],
        performances: [],
        additionalPerformances: [],
        specialPerformances: [],
        performancesClosures: [],
        talents: [],
        mediumTags: [],
        styleTags: [],
        audienceTags: [],
        geoTags: [],
        images: [],
        version: 0,
        createdDate: null
      }
    case entityConstants.ENTITY_TYPE_EVENT_SERIES:
      return {
        id: null,
        status: entityConstants.ACTIVE_STATUS,
        validStatuses: [entityConstants.ACTIVE_STATUS],
        name: '',
        eventSeriesType: '',
        occurrence: '',
        summary: '',
        description: RichTextEditor.createEmptyValue(),
        images: [],
        links: [],
        descriptionCredit: '',
        weSay: '',
        version: 0,
        createdDate: null
      }
    default:
      throw new Error(`entityType ${entityType} not supported`)
  }
}
