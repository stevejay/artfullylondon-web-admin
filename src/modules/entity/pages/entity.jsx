import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'

import { Error } from '_src/modules/error'
import FadeTransition from '_src/components/transition/fade'
import BoxesLoader from '_src/components/loader/boxes'
import BasicSection from '_src/components/section/basic'
import * as entitiesPropTypes from '_src/entities/prop-types'
import * as entityConstants from '_src/constants/entity'
import * as entityActions from '_src/modules/entity/actions'
import { selectors as entitySelectors } from '_src/modules/entity/reducers'

export class EntityPage extends React.Component {
  componentWillMount () {
    this._getOrResetEntity(this.props)
  }
  componentWillReceiveProps (nextProps) {
    if (nextProps.location !== this.props.location) {
      this._getOrResetEntity(nextProps)
    }
  }
  _getOrResetEntity ({ entityType, entityId }) {
    this.props.dispatch(
      entityId
        ? entityActions.getEntity(entityType, entityId)
        : entityActions.resetEntityForCreate(entityType)
    )
  }
  render () {
    const {
      entityId,
      entity,
      getInProgress,
      getFailed,
      component: Component
    } = this.props

    if (getFailed) {
      return <BasicSection><Error /></BasicSection>
    }

    if (getInProgress || !entity) {
      return <BasicSection><BoxesLoader /></BasicSection>
    }

    return (
      <FadeTransition in appear mountOnEnter unmountOnExit>
        <BasicSection>
          <Component entity={entity} isEdit={!!entityId} />
        </BasicSection>
      </FadeTransition>
    )
  }
}

EntityPage.propTypes = {
  entityType: PropTypes.oneOf(entityConstants.EDITABLE_ENTITY_TYPES).isRequired,
  entityId: PropTypes.string,
  entity: entitiesPropTypes.EDITABLE_ENTITY,
  getInProgress: PropTypes.bool.isRequired,
  getFailed: PropTypes.bool.isRequired,
  component: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  match: PropTypes.object.isRequired,
  location: PropTypes.object.isRequired
}

export default withRouter(
  connect(
    /* istanbul ignore next */
    (state, ownProps) => ({
      // TODO create selector for match param
      entityId: ownProps.match.params[0],
      entity: entitySelectors.entity(state),
      getInProgress: entitySelectors.gettingEntity(state),
      getFailed: entitySelectors.failedToGetEntity(state)
    })
  )(EntityPage)
)
