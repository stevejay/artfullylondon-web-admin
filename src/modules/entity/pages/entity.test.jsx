import React from 'react'
import _ from 'lodash'
import log from 'loglevel'

import { FullTalent } from '_src/domain/talent'
import { EntityPage } from './entity'
import * as entityActions from '../actions'
import { actions as tagActions } from '_src/modules/tag'

class SomeComponent extends React.Component {
  render () {
    return <div id='component' />
  }
}

it('should render correctly when an error was raised', () => {
  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError
      setHasError={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when an error was raised', () => {
  const setHasError = jest.fn()
  log.error = jest.fn()

  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={setHasError}
    />
  )

  wrapper.instance().componentDidCatch(new Error('deliberately thrown'), {})

  expect(setHasError).toHaveBeenCalledWith(true)
  expect(log.error).toHaveBeenCalled()
})

it('should render correctly when get entity failed', () => {
  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when get is in progress', () => {
  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when getting tags', () => {
  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should render correctly when has an entity', () => {
  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={new FullTalent({ id: 'some-id' })}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should trigger getting the entity on construction when viewing an entity', () => {
  const dispatch = jest.fn()

  shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={dispatch}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(dispatch).toHaveBeenCalledWith(
    entityActions.getEntity(entityType.TALENT, 'some-id')
  )

  expect(dispatch).not.toHaveBeenCalledWith(tagActions.getTags())
})

it('should trigger getting the entity on construction when editing an entity', () => {
  const dispatch = jest.fn()

  shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing={false}
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={dispatch}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(dispatch).toHaveBeenCalledWith(
    entityActions.getEntity(entityType.TALENT, 'some-id')
  )

  expect(dispatch).toHaveBeenCalledWith(tagActions.getTags())
})

it('should trigger getting the entity on construction when creating an entity', () => {
  const dispatch = jest.fn()

  shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing={false}
      entityId={null}
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={dispatch}
      match={{}}
      location={{}}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  expect(dispatch).toHaveBeenCalledWith(
    entityActions.resetEntityForCreate(entityType.TALENT)
  )

  expect(dispatch).toHaveBeenCalledWith(tagActions.getTags())
})

it('should not trigger getting the entity when props change but the same entity is specified', () => {
  const dispatch = jest.fn()
  const location = { path: '/first' }

  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={dispatch}
      match={{}}
      location={location}
      history={{}}
      hasError={false}
      setHasError={_.noop}
    />
  )

  wrapper.instance().componentWillReceiveProps({ location })

  expect(dispatch).toHaveBeenCalledTimes(1)
})

it('should trigger getting the entity when props change and a different entity is specified', () => {
  const dispatch = jest.fn()
  const setHasError = jest.fn()

  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing
      entityId='some-id'
      entity={null}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={dispatch}
      match={{}}
      location={{ path: '/first' }}
      history={{}}
      hasError={false}
      setHasError={setHasError}
    />
  )

  wrapper.instance().componentWillReceiveProps({
    location: { path: '/second' },
    entityType: entityType.TALENT,
    entityId: 'some-other-id',
    dispatch
  })

  expect(setHasError).toHaveBeenCalledWith(false)

  expect(dispatch).toHaveBeenCalledWith(
    entityActions.getEntity(entityType.TALENT, 'some-id')
  )
})

it('should handle cancelling the entity edit', () => {
  const history = { goBack: jest.fn() }
  const event = { preventDefault: jest.fn() }

  const wrapper = shallow(
    <EntityPage
      entityType={entityType.TALENT}
      viewing={false}
      entityId='some-id'
      entity={new FullTalent({ id: 'some-id' })}
      getInProgress={false}
      getFailed={false}
      gettingTags={false}
      component={SomeComponent}
      dispatch={_.noop}
      match={{}}
      location={{}}
      history={history}
      hasError={false}
      setHasError={_.noop}
    />
  )

  wrapper.find(SomeComponent).prop('onCancel')(event)

  expect(event.preventDefault).toHaveBeenCalled()
  expect(history.goBack).toHaveBeenCalled()
})
