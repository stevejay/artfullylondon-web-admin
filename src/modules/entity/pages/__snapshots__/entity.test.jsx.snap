// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when an error was raised 1`] = `
<BasicSection>
  <Error
    type="DEFAULT"
  />
</BasicSection>
`;

exports[`should render correctly when get entity failed 1`] = `
<BasicSection>
  <Error
    type="DEFAULT"
  />
</BasicSection>
`;

exports[`should render correctly when get is in progress 1`] = `
<BasicSection>
  <BoxesLoader
    size="large"
    type="default"
  />
</BasicSection>
`;

exports[`should render correctly when getting tags 1`] = `
<BasicSection>
  <BoxesLoader
    size="large"
    type="default"
  />
</BasicSection>
`;

exports[`should render correctly when has an entity 1`] = `
<FadeTransition
  appear={true}
  in={true}
  mountOnEnter={true}
  unmountOnExit={true}
>
  <BasicSection>
    <SomeComponent
      entity={
        FullTalent {
          "id": "some-id",
        }
      }
      isEdit={true}
      onCancel={[Function]}
    />
  </BasicSection>
</FadeTransition>
`;
